### ISSUE TITLE
[Bug] SEVERE downstream task performance degradation compared to uncompiled model

### ISSUE THREAD

**Issue by 0xLienid** [mlc-ai/mlc-llm#2499]
## 🐛 Bug

Compiled models, even with no quantization, perform MUCH worse than their uncompiled equivalent on downstream tasks (measured by MMLU). This is extra true for Llama where it's so bad I'm convinced there's an error in the model implementation as Llama-2 and Llama-3 score essentially identically.

### Llama-3 8B
HF Transformers (sampled category scores):
- Average accuracy 0.280 - abstract_algebra
- Average accuracy 0.652 - anatomy
- Average accuracy 0.737 - astronomy
- Average accuracy 0.720 - business_ethics
- Average accuracy 0.751 - clinical_knowledge
- Average accuracy 0.799 - college_biology

MLC (loaded, compiled directly from HF): 
- Average accuracy 0.220 - abstract_algebra
- Average accuracy 0.185 - anatomy
- Average accuracy 0.178 - astronomy
- Average accuracy 0.300 - business_ethics
- Average accuracy 0.215 - clinical_knowledge
- Average accuracy 0.257 - college_biology

### Llama-2 7B MLC comparison
- Average accuracy 0.220 - abstract_algebra
- Average accuracy 0.185 - anatomy
- Average accuracy 0.178 - astronomy
- Average accuracy 0.300 - business_ethics
- Average accuracy 0.215 - clinical_knowledge
- Average accuracy 0.257 - college_biology 

### Mistral v0.3
HF Transformers
- Average accuracy 0.290 - abstract_algebra
- Average accuracy 0.600 - anatomy
- Average accuracy 0.645 - astronomy
- Average accuracy 0.610 - business_ethics
- Average accuracy 0.679 - clinical_knowledge
- Average accuracy 0.715 - college_biology 

MLC
- Average accuracy 0.280 - abstract_algebra
- Average accuracy 0.430 - anatomy
- Average accuracy 0.480 - astronomy
- Average accuracy 0.530 - business_ethics
Average accuracy 0.540 - clinical_knowledge
Average accuracy 0.528 - college_biology 

## To Reproduce

1. Build MLC from source, modifying the limit on number of logprobs to be higher
2. Create MLCEngine object that loads model from HF
3. Run it through MMLU
4. Try asking MMLU questions directly with the model via chat to confirm it's not an MMLU implementation issue

## Expected behavior

Model does not have 33-50% the performance of an uncompiled model

## Environment

 - Platform (e.g. WebGPU/Vulkan/IOS/Android/CUDA): CUDA and Metal
 - Operating system (e.g. Ubuntu/Windows/MacOS/...): Ubuntu and MacOS
 - Device (e.g. iPhone 12 Pro, PC+RTX 3090, ...): 1xA100, 2xA100, 1x4090, MBP
 - How you installed MLC-LLM (`conda`, source): source
 - How you installed TVM-Unity (`pip`, source): pip
 - Python version (e.g. 3.10): 3.12.3
 - CUDA/cuDNN version (if applicable): 12.4

**Comment by MasterJH5574** [comment]
Hi @0xLienid thanks for bringing up this issue! Do you mind letting us know at which commit your codebase is? We fixed some precision issue last week in #2449 and would like to know if your codebase is older or newer than that.

**Comment by MasterJH5574** [comment]
Thank you @0xLienid for sharing the info. We'll look into this issue. Meanwhile, to help better reproduce exactly the same MLC results, would you like tell us how you “run MLC through MMLU”?

**Comment by 0xLienid** [comment]
1. increase constraint on max number of logprobs to 150
2. build MMLU prompt using the process from hendrycks/test
3. determine selection of A, B, C, D via 
```
        response = model_engine.chat.completions.create(
            messages=[{"role": "user", "content": prompt}],
            model="mlc-ai/Llama-3-70B-Instruct-q3f16_1-MLC",
            stream=False,
            max_tokens=1,
            temperature=1.0,
            logprobs=True,
            top_logprobs=150
        )

        token_logprobs = {}
        logprobs = response.choices[0].logprobs.content[0].top_logprobs
        for logprob in logprobs:
            token_logprobs[logprob.token] = logprob.logprob

        abcd_logprobs = {}
        for choice in MMLU_CHOICES:
            if choice in token_logprobs.keys():
                abcd_logprobs[choice] = token_logprobs[choice]
            else:
                abcd_logprobs[choice] = -100

        pred = {0: "A", 1: "B", 2: "C", 3: "D"}[
            np.argmax(list(abcd_logprobs.values()))]
```
4. determine percentage correct

this is based on the expectation that logprobs and logits have the same ordering which should be true, correct?

**Comment by MasterJH5574** [comment]
@0xLienid Thank you so much for providing the detailed script, which helped triage a lot!

A quick note is that for chat completion, we need to provide a system prompt, in order to force the model output only `A`, `B`, `C` or `D`. Below is an example for the astronomy MMLU tests.
```python
    prompt = f"""
The following are multiple choice questions (with answers) about astronomy.
{question}
A. {choice_A}
B. {choice_B}
C. {choice_C}
D. {choice_D}
Answer:
""".strip()
    system_prompt = 'You are a helpful assistant. Answer the multiple choice question and always respond only with "A", "B", "C", or "D".'
    response = engine.chat.completions.create(
        messages=[
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": prompt},
        ],
        stream=False,
        max_tokens=1,
        temperature=1.0,
        logprobs=True,
        top_logprobs=20,
    )
```

If we do not provide the system prompt, the default system prompt in MLC for Llama3 is `"You are a helpful, respectful and honest assistant."`. In this case, the token with highest logprob is `"The"` when running unquantized Llama3:
```
logprobs = [
    TopLogProbs(token='The', logprob=-0.229968, bytes=[84, 104, 101]),
    TopLogProbs(token='Based', logprob=-2.22997, bytes=[66, 97, 115, 101, 100]),
    TopLogProbs(token='I', logprob=-2.32372, bytes=[73]),
    TopLogProbs(token='!', logprob=-87.3365, bytes=[33]),
    ...
]
abcd_logprobs = {'A': -100, 'B': -100, 'C': -100, 'D': -100}
```

As you can see, the `abcd_logprobs` is full of `-100`, and after argmax, the answer will become always `"A"`.

So the takeaway here is we'll need to provide the system prompt when using chat completion. Besides, when cross comparing the MLC chat completion results with HF Transformers results, we would want to align the prompt (essentially the input token ids), given the model can be sensitive to prompts.

I quickly checked the unquantized Llama3 accuracy on the astronomy test set for both MLC and HF with aligned prompts, and both sides answer 103 questions correctly out of 152 in total (so the accuracy is 0.6777).

So would you mind checking the system prompt and try again? Thanks again for bringing this up, and let us know if there is any further problems :-)

**Comment by 0xLienid** [comment]
I mean, that is in essence cheating the benchmark. I'm not giving any such "always respond with only 'A', 'B', 'C', or 'D'" instructions to the HF model. When I override the system prompt in MLC so that it is `"The following are multiple choice questions (with answers) about {}.\n\n".format(format_subject(subject))` rather than the base one, I get the same performance.

**Comment by MasterJH5574** [comment]
Thank you @0xLienid for the follow-up. I understand your point that for benchmarking we may want to avoid any kinds of hint. In this case I would like to note the difference between chat completions and raw-text generations.

* In our discussions above we used `model_engine.chat.completions.create(...)`, which involves the [conversation template](https://github.com/meta-llama/llama3?tab=readme-ov-file#instruction-tuned-models) that the Llama3 instruct model is specially fined tuned for. This means that the model will behave more like a chat bot, and that's the reason it sometimes does not speak out the answer directly, unless we give clear instruction to the model saying that “you need to output A/B/C/D directly”.

* On the other hand, if the chat bot behavior is not what we desire here (say we don't want to give that instruction, and want the model to answer the question directly), it is more appropriate to use the raw-text generation via `model_engine.completions.create(prompt=prompt, ...)`. This interface directly tokenizes the input prompt, without any conversation template process. And in this case the model will behave less like the chat bot.

I hope I elaborated the difference between chat completion and raw-text generation clear enough (you're welcome to point out anything that is confusing to you). To summarize, for our purpose here we can use `model_engine.completions.create`. Through this interface, what users pass in will be exactly what the model receives. And here is an example we can use in the benchmark (note that we don't provide anything else but only the question and choices, and I guess this is also what you only provide when running HF Transformers):

```python
prompt = f"""
{question}
A. {choice_A}
B. {choice_B}
C. {choice_C}
D. {choice_D}
Answer:
""".strip()
response = model_engine.completions.create(
    prompt=prompt,
    stream=False,
    max_tokens=1,
    temperature=1.0,
    logprobs=True,
    top_logprobs=150,
)

token_logprobs = {}
logprobs = response.choices[0].logprobs.content[0].top_logprobs
for logprob in logprobs:
    token_logprobs[logprob.token] = logprob.logprob
...
```

**Comment by 0xLienid** [comment]
did that give the expected performance for you? that gives the exact same performance for me.

**Comment by MasterJH5574** [comment]
> did that give the expected performance for you?

@0xLienid Yes that's right. Here are the results:

Llama-3-8B-Instruct-q0f16-MLC:
- abstract_algebra accuracy: 29 / 100 = 0.29
- anatomy accuracy: 87 / 135 = 0.6444444444444445
- astronomy accuracy: 110 / 152 = 0.7236842105263158
- business_ethics accuracy: 67 / 100 = 0.67
- clinical_knowledge accuracy: 194 / 265 = 0.7320754716981132
- college_biology accuracy: 114 / 144 = 0.7916666666666666

Mistral-7B-Instruct-v0.3-q0f16-MLC:
- abstract_algebra accuracy: 34 / 100 = 0.34
- anatomy accuracy: 74 / 135 = 0.5481481481481482
- astronomy accuracy: 101 / 152 = 0.6644736842105263
- business_ethics accuracy: 63 / 100 = 0.63
- clinical_knowledge accuracy: 183 / 265 = 0.690566037735849
- college_biology accuracy: 100 / 144 = 0.6944444444444444

And here is the Python file to reproduce:
```python
import csv
from pathlib import Path

import numpy as np

from mlc_llm import MLCEngine

subjects = [
    "abstract_algebra",
    "anatomy",
    "astronomy",
    "business_ethics",
    "clinical_knowledge",
    "college_biology",
]


def run_tests():
    mmlu_test_path = Path("workspace/mmlu_data/test/")
    engine = MLCEngine("HF://mlc-ai/Llama-3-8B-Instruct-q0f16-MLC")
    # engine = MLCEngine("HF://mlc-ai/Llama-2-7b-chat-hf-q0f16-MLC")
    # engine = MLCEngine("HF://mlc-ai/Mistral-7B-Instruct-v0.3-q0f16-MLC")
    for subject in subjects:
        with open(mmlu_test_path / f"{subject}_test.csv", newline="") as csvfile:
            tests = list(csv.reader(csvfile, delimiter=",", quotechar='"'))
            assert all(len(test) == 6 for test in tests)

        num_correct = 0
        error_list = []
        for i, test in enumerate(tests):
            prompt = f"""
{test[0]}
A. {test[1]}
B. {test[2]}
C. {test[3]}
D. {test[4]}
Answer:
""".strip()
            response = engine.completions.create(
                prompt=prompt,
                stream=False,
                max_tokens=1,
                temperature=1.0,
                logprobs=True,
                top_logprobs=20,
            )

            token_logprobs = {}
            logprobs = response.choices[0].logprobs.content[0].top_logprobs
            for logprob in logprobs:
                if logprob.token not in token_logprobs:
                    token_logprobs[logprob.token] = logprob.logprob

            abcd_logprobs = {}
            for choice in ["A", "B", "C", "D"]:
                if choice in token_logprobs.keys():
                    abcd_logprobs[choice] = token_logprobs[choice]
                else:
                    abcd_logprobs[choice] = -100

            pred = {0: "A", 1: "B", 2: "C", 3: "D"}[np.argmax(list(abcd_logprobs.values()))]
            if pred == test[5]:
                num_correct += 1
            else:
                error_list.append(i)

        print(f"{subject} accuracy: {num_correct} / {len(tests)} = {num_correct / len(tests)}")


if __name__ == "__main__":
    run_tests()
```

**Comment by 0xLienid** [comment]
ok, I am matching performance now. thank you for all the help. one note, I would make it more clear in the documentation what approaches to sending prompts to a model exist. right now it's not at all obvious there is anything but `engine.chat.completions.create` unless you dig into the code.

**Comment by MasterJH5574** [comment]
Glad to hear that it's matching now. And thank you for the great suggestion! We will iterate the document to better reflect the raw-text completion interface.
